project/Component/NefSwiftPlayground/SwiftPlayground.swift

```diff
@@ -0,0 +1,170 @@
+            let resolvePackageIO = shell.resolve(packagePath: packagePath, buildPath: buildPath).mapLeft { _ in SwiftPlaygroundError.checkout }
+
+            return writePackageIO.followedBy(resolvePackageIO)
+        }
+    }
+    
+    private func repositories(checkoutPath: String) -> EnvIO<FileSystem, SwiftPlaygroundError, [String]> {
+        EnvIO { fileSystem in
+            fileSystem.items(atPath: checkoutPath)
+                      .mapLeft { _ in .checkout }
+                      .map { repos in repos.filter { repo in !repo.filename.contains("swift-") } }
+        }
+    }
+    
+    private func swiftLibraryModules(in repos: [String], excludes: [PlaygroundExcludeItem]) -> EnvIO<PlaygroundShell, SwiftPlaygroundError, [Module]> {
+        func modules(in data: Data) -> [Module] {
+            guard let package = try? JSONDecoder().decode(Package.self, from: data) else { return [] }
+            
+            let modules = package.targets.filter { module in module.type == .library &&
+                                                             module.moduleType == .swift &&
+                                                            !excludes.contains(.module(name: module.name)) }
+            
+            return Module.moduleNameAndSourcesTraversal.modify(modules) { (name, sources) in
+                (name, sources.filter { file in !excludes.contains(.file(name: file.filename, module: name)) })
+            }
+        }
+        
+        func linkPathForSources(in modules: [Module]) -> EnvIO<PlaygroundShell, PlaygroundShellError, [Module]> {
+            EnvIO { shell in
+                modules.traverse { (module: Module) in
+                    let sourcesIO = module.sources.map { ($0, module.path) }.traverse(shell.linkPath)^
+                    let moduleIO  = sourcesIO.map { sources in
+                        Module.sourcesLens.modify(module, { _ in sources})
+                    }^
+                    
+                    return moduleIO
+                }
+            }
+        }
+        
+        return EnvIO { shell in
+            repos.parFlatTraverse { repositoryPath in
+                shell.describe(repositoryPath: repositoryPath)
+                     .map(modules)^
+                     .flatMap { modules in linkPathForSources(in: modules).provide(shell) }^
+                     .mapLeft { _ in .modules(repos) }
+            }.flatMap { modules in
+                modules.count > 0 ? IO.pure(modules)^
+                                  : IO.raiseError(.modules(repos))^
+            }^
+        }
+    }
+    
+    private func buildPlaygroundBook(modules: [Module], playgroundPath: String) -> EnvIO<FileSystem, SwiftPlaygroundError, Void> {
+        EnvIO { system in
+            PlaygroundBook(name: "nef", path: playgroundPath)
+                .build(modules: modules)
+                .provide(system)
+                .mapLeft { e in .playgroundBook(info: e.description) }
+        }
```

Comment:
What about:

```swift
PlaygroundBook(name: "nef", path: playgroundPath)
                .build(modules: modules)
                .mapError { e in .playgroundBook(info: e.description) }
```

No need to wrap it in an `EnvIO`, it is already an `EnvIO`, only need to transform the error type.

Reply:
You are right!